From babbc89f628be4167f923af71682b6b7e2810882 Mon Sep 17 00:00:00 2001
From: kikairoya <kikairoya@gmail.com>
Date: Sun, 6 Sep 2026 13:15:09 +0900
Subject: [PATCH 44/63] [Clang][Cygwin] Support -mno-use-libstdc-wrappers

Cygwin provides a mechanism to share overridden `operator new`, etc., in a DLL from a host EXE or other DLLs, by
cooperating with the linker's `--wrap` option.
GCC has a Cygwin-specific option, `-mno-use-libstdc-wrappers`, to suppress such `--wrap`s.
This patch implements it to align with GCC.
---
 clang/include/clang/Options/Options.td |  2 ++
 clang/lib/Driver/ToolChains/Cygwin.cpp | 41 +++++++++++++++-----------
 clang/test/Driver/cygwin.cpp           |  4 +++
 3 files changed, 29 insertions(+), 18 deletions(-)

diff --git a/clang/include/clang/Options/Options.td b/clang/include/clang/Options/Options.td
index c4c49df73c15ae2bc267d5217987f02a1663f242..30551968a5442f16eda328b6eda1688d3665feea 100644
--- a/clang/include/clang/Options/Options.td
+++ b/clang/include/clang/Options/Options.td
@@ -5627,6 +5627,8 @@ def mconsole : Joined<["-"], "mconsole">, Group<m_Group>;
 def mwindows : Joined<["-"], "mwindows">, Group<m_Group>;
 def mdll : Joined<["-"], "mdll">, Group<m_Group>;
 def municode : Joined<["-"], "municode">, Group<m_Group>;
+def muse_libstdc_wrappers : Joined<["-"], "muse-libstdc-wrappers">, Group<m_Group>;
+def mno_use_libstdc_wrappers : Joined<["-"], "mno-use-libstdc-wrappers">, Group<m_Group>;
 def mthreads : Joined<["-"], "mthreads">, Group<m_Group>;
 def marm64x : Joined<["-"], "marm64x">, Group<m_Group>,
   Visibility<[ClangOption, CLOption]>,
diff --git a/clang/lib/Driver/ToolChains/Cygwin.cpp b/clang/lib/Driver/ToolChains/Cygwin.cpp
index 4b64557eff495b7aa8fc53ae3649ba9ea4545fc7..7af7c806d4476bcf521334c9c5d2ad2c3eb3630c 100644
--- a/clang/lib/Driver/ToolChains/Cygwin.cpp
+++ b/clang/lib/Driver/ToolChains/Cygwin.cpp
@@ -183,24 +183,29 @@ void cygwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     CmdArgs.push_back("console");
   }
 
-  if (ToolChain.getTriple().isArch32Bit()) {
-    CmdArgs.push_back("--wrap=_Znwj");
-    CmdArgs.push_back("--wrap=_Znaj");
-    CmdArgs.push_back("--wrap=_ZdlPv");
-    CmdArgs.push_back("--wrap=_ZdaPv");
-    CmdArgs.push_back("--wrap=_ZnwjRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZnajRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
-  } else {
-    CmdArgs.push_back("--wrap=_Znwm");
-    CmdArgs.push_back("--wrap=_Znam");
-    CmdArgs.push_back("--wrap=_ZdlPv");
-    CmdArgs.push_back("--wrap=_ZdaPv");
-    CmdArgs.push_back("--wrap=_ZnwmRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZnamRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+  Arg *WrapArg = Args.getLastArg(options::OPT_mno_use_libstdc_wrappers,
+                                 options::OPT_muse_libstdc_wrappers);
+
+  if (!WrapArg || WrapArg->getOption().matches(options::OPT_muse_libstdc_wrappers)) {
+    if (ToolChain.getTriple().isArch32Bit()) {
+      CmdArgs.push_back("--wrap=_Znwj");
+      CmdArgs.push_back("--wrap=_Znaj");
+      CmdArgs.push_back("--wrap=_ZdlPv");
+      CmdArgs.push_back("--wrap=_ZdaPv");
+      CmdArgs.push_back("--wrap=_ZnwjRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZnajRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+    } else {
+      CmdArgs.push_back("--wrap=_Znwm");
+      CmdArgs.push_back("--wrap=_Znam");
+      CmdArgs.push_back("--wrap=_ZdlPv");
+      CmdArgs.push_back("--wrap=_ZdaPv");
+      CmdArgs.push_back("--wrap=_ZnwmRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZnamRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
+      CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+    }
   }
 
   if (Args.hasArg(options::OPT_mdll))
diff --git a/clang/test/Driver/cygwin.cpp b/clang/test/Driver/cygwin.cpp
index 66fa9fbcfc1e0ad5557326587b44307b527062f8..7c00ac0e1219b2d4c7b973b2bcb0e79c120b8728 100644
--- a/clang/test/Driver/cygwin.cpp
+++ b/clang/test/Driver/cygwin.cpp
@@ -167,6 +167,10 @@
 // CHECK-64-WRAP-SAME: "--wrap=_ZdlPvRKSt9nothrow_t"
 // CHECK-64-WRAP-SAME: "--wrap=_ZdaPvRKSt9nothrow_t"
 
+// RUN: %clang -### %s --target=x86_64-pc-cygwin --sysroot=%S/Inputs/basic_cygwin_tree \
+// RUN:  -mno-use-libstdc-wrappers 2>&1 | FileCheck --check-prefix=CHECK-NOWRAP %s
+// CHECK-NOWRAP-NOT: --wrap
+
 // RUN: %clang -### %s --target=x86_64-pc-cygwin --sysroot=%S/Inputs/basic_cygwin_tree \
 // RUN:                -resource-dir=%S/Inputs/resource_dir -rtlib=compiler-rt \
 // RUN:   2>&1 | FileCheck --check-prefix=CHECK-64-RTLIB %s
-- 
2.55.0

